fedora/clients/fasjson.py

```diff
@@ -101,3 +101,51 @@ async def get_users_by_matrix_id(self, matrix_id: str) -> dict | str:
             )
 
         return searchresult[0]
+
+    async def get_user_groups(self, username, params=None):
+        """
+        Retrieves all groups a user belongs to, including membership type (member or sponsor).
+
+        Args:
+            username (str): Username of the user to query.
+            params (dict, optional): Additional query parameters to include in the request.
+
+        Returns:
+            List[dict] or None:
+                - A list of dictionaries containing group information:
+                    - groupname (str): Name of the group.
+                    - membership_type (str): "member" or "sponsor" depending on the user's role.
+                - None (if the user is not found or there's an error).
+
+        Raises:
+            InfoGatherError: If there's an error fetching data from Fasjson or if the user is not found (404).
+        """
+
+        try:
+            response = await self._get("/".join(["users", username, "groups"]), params=params)
+            user_groups = response.json().get("groups", [])
+
+            group_details = []
+            for groupname in user_groups:
+                # Fix - i realized at some point that get group membership, just returns members/sponsors in a group
+                sponsors = await self.get_group_membership(groupname, "sponsors", params=params)
+
+            if sponsors is not None:
+                if username in sponsors:
+                    membership_type = "sponsor"
+                else:
+                    membership_type = "member"
+            else:
+                # Handle the case when sponsors is None (no sponsors for the group)
+                membership_type = "member"
+
+            group_details.append({"groupname": groupname, "membership_type": membership_type})
+
+            return group_details
+
+        except NoResult as e:
```

Comment:
Based on the error message, this exception could happen on line 125. I recommend moving this except block up to line  126 so the try/except block cover the smallest amount of code necessary.

The advantage of keeping the try/except blocks as small as possible is that it's clearer what could raise the exceptions and you won't accidentally handle exceptions in the subsequent code in a way you didn't intend. For example, if someone were to add another `self._get` call in this function and neglected to handle a `NoResult`, or if someone changed `self.get_group_membership` to not handle `NoResult`, this would catch it and return an error saying the username doesn't exist, even though it might.